hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/upgrade/TestHDDSUpgrade.java

```diff
@@ -305,4 +405,655 @@ public void testFinalizationFromInitialVersionToLatestVersion()
     // Verify that new pipeline can be created with upgraded datanodes.
     testPostUpgradePipelineCreation();
   }
+
+  /*
+   * All the subsequent tests here are failure cases. Some of the tests below
+   * could simultaneously fail one or more nodes at specific execution points
+   * and in different thread contexts.
+   * Upgrade path key execution points are defined in
+   * UpgradeFinalizer:UpgradeTestInjectionPoints.
+   */
+
+  /*
+   * Helper function to inject SCM failure and a SCM restart at a given
+   * execution point during SCM-Upgrade.
+   *
+   * Injects Failure in  : SCM
+   * Executing-Thread-Context : SCM-Upgrade
+   */
+  private Boolean injectSCMFailureDuringSCMUpgrade()
+      throws InterruptedException, TimeoutException, AuthenticationException,
+      IOException {
+    // For some tests this could get called in a different thread context.
+    // We need to guard concurrent updates to the cluster.
+    synchronized(cluster) {
+      cluster.restartStorageContainerManager(true);
+      loadSCMState();
+    }
+    // The ongoing current SCM Upgrade is getting aborted at this point. We
+    // need to schedule a new SCM Upgrade on a different thread context.
+    Thread t = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          loadSCMState();
+          scm.finalizeUpgrade("xyz");
+        } catch (IOException e) {
+          e.printStackTrace();
+          testPassed.set(false);
+        }
+      }
+    });
+    t.start();
+    return true;
+  }
+
+  /*
+   * Helper function to inject DataNode failures and DataNode restarts at a
+   * given execution point during SCM-Upgrade. Please note that it fails all
+   * the DataNodes in the cluster and is part of test cases that simulate
+   * multi-node failure at specific code-execution points during SCM Upgrade.
+   * Please note that this helper function should be called in the thread
+   * context of an SCM-Upgrade only. The return value has a significance that
+   * it does not abort the currently ongoing SCM upgrade. because this
+   * failure injection does not fail the SCM node and only impacts datanodes,
+   *  we do not need to schedule another scm-finalize-upgrade here.
+   *
+   * Injects Failure in  : All the DataNodes
+   * Executing-Thread-Context : SCM-Upgrade
+   */
+  private Boolean injectDataNodeFailureDuringSCMUpgrade() {
+    try {
+      // Work on a Copy of current set of DataNodes to avoid
+      // running into tricky situations.
+      List<HddsDatanodeService> currentDataNodes =
+          new ArrayList<>(cluster.getHddsDatanodes());
+      for (HddsDatanodeService ds: currentDataNodes) {
+        DatanodeDetails dn = ds.getDatanodeDetails();
+        cluster.restartHddsDatanode(dn, false);
```

Comment:
Apart from immediate restart of the DN, can we also look at a case where the DN was brought down (or already down) during finalization, and can join late after finalization is completed?

Do we have a test case where a new node (DN MLV = DN SLV > SCM MLV) is registering during finalization or when SCM is in pre-Finalize?

Reply:
Not these will be new test cases. We can keep on adding new test cases like these.